crates/hyperdrive-math/src/long/close.rs

```diff
@@ -53,10 +53,54 @@ impl State {
 
         Ok(flat + curve)
     }
+
+    /// Calculates the market value in shares of a long position using the equation:
+    /// market_estimate = trading_proceeds - flat_fees_paid - curve_fees_paid
+    ///
+    /// trading_proceeds = flat_value + curve_value
+    ///                  = dy * (1 - t) + dy * t * p
+    ///                  = dy * ((1 - t) + (t * p))
+    /// flat_value       = dy * (1 - t)
+    /// curve_bonds      = dy * t
+    /// curve_value      = curve_bonds * p
+    ///                  = dy * t * p
+    /// flat_fees_paid   = flat_value * flat_fee
+    ///                  = dy * (1 - t) * flat_fee
+    /// curve_fees_paid  = (curve_bonds - curve_value) * curve_fee
+    ///                  = dy * t * (1 - p) * curve_fee
+    ///
+    /// dy = bond_amount
+    /// p  = spot_price
+    /// t  = time_remaining
+    pub fn calculate_value_long<F: Into<FixedPoint>>(
+        &self,
+        bond_amount: F,
+        maturity_time: U256,
+        current_time: U256,
+    ) -> Result<FixedPoint> {
+        let bond_amount = bond_amount.into();
+
+        let spot_price = self.calculate_spot_price()?;
+
+        // get the time remaining
+        let time_remaining = self.calculate_normalized_time_remaining(maturity_time, current_time);
+
+        let flat_value = bond_amount * (fixed!(1e18) - time_remaining);
+        let curve_bonds = bond_amount * time_remaining;
+        let curve_value = curve_bonds * spot_price;
+
+        let trading_proceeds = flat_value + curve_value;
+        let flat_fees_paid = flat_value * self.config.fees.flat.into();
+        let curve_fees_paid = (curve_bonds - curve_value) * self.config.fees.curve.into();
+
+        Ok(trading_proceeds - flat_fees_paid - curve_fees_paid)
```

Comment:
Same as above, verify that these numbers will work out. You could compute `fees_paid = flat_fees_paid + curve_fees_paid` and then check `trading_proceeds >= fees_paid`. Or do two checks, one for each subtraction.

Reply:
Updated.